contracts/pair_stable/src/math.rs

```diff
@@ -73,122 +18,199 @@ pub fn calc_offer_amount(
 /// A * sum(x_i) * n**n + D = A * D * n**n + D**(n+1) / (n**n * prod(x_i))
 ///
 /// ## Params
-/// * **leverage** is an object of type [`u128`].
-///
-/// * **amount_a** is an object of type [`u128`].
-///
-/// * **amount_b** is an object of type [`u128`].
-pub fn compute_d(leverage: u64, amount_a: u128, amount_b: u128) -> Option<u128> {
-    let amount_a_times_coins =
-        checked_u8_mul(&U256::from(amount_a), N_COINS)?.checked_add(U256::one())?;
-    let amount_b_times_coins =
-        checked_u8_mul(&U256::from(amount_b), N_COINS)?.checked_add(U256::one())?;
-    let sum_x = amount_a.checked_add(amount_b)?; // sum(x_i), a.k.a S
-    if sum_x == 0 {
-        Some(0)
-    } else {
-        let mut d_previous: U256;
-        let mut d: U256 = sum_x.into();
+/// * **amp** is an object of type [`Uint64`].
+///
+/// * **pools** is a vector with values of type [`Decimal256`].
+///
+/// * **greatest_precision** object of type [`u8`].
+pub(crate) fn compute_d(
+    amp: Uint64,
+    pools: &[Decimal256],
+    greatest_precision: u8,
+) -> StdResult<Decimal256> {
+    if pools.iter().any(|pool| pool.is_zero()) {
+        return Ok(Decimal256::zero());
+    }
+    let sum_x = pools.iter().fold(Decimal256::zero(), |acc, x| acc + (*x));
 
-        // Newton's method to approximate D
+    if sum_x.is_zero() {
+        Ok(Decimal256::zero())
+    } else {
+        let n_coins = pools.len() as u8;
+        let ann = Decimal256::from_integer(amp.checked_mul(n_coins.into())?.u64() / AMP_PRECISION);
+        let n_coins = Uint64::from(n_coins);
+        let mut d = sum_x;
+        let ann_sum_x = ann * sum_x;
         for _ in 0..ITERATIONS {
-            let mut d_product = d;
-            d_product = d_product
-                .checked_mul(d)?
-                .checked_div(amount_a_times_coins)?;
-            d_product = d_product
-                .checked_mul(d)?
-                .checked_div(amount_b_times_coins)?;
-            d_previous = d;
-            // d = (leverage * sum_x + d_p * n_coins) * d / ((leverage - 1) * d + (n_coins + 1) * d_p);
-            d = calculate_step(&d, leverage, sum_x, &d_product)?;
-            // Equality with the precision of 1
-            if d == d_previous {
-                break;
+            // loop: D_P = D_P * D / (_x * N_COINS)
+            let d_p = pools
+                .iter()
+                .try_fold::<_, _, StdResult<_>>(d, |acc, pool| {
+                    let denominator = pool.atomics().checked_mul(n_coins.into())?;
+                    acc.checked_multiply_ratio(d, Decimal256::new(denominator))
+                })?;
+            let d_prev = d;
+            d = (ann_sum_x + d_p * Decimal256::from_integer(n_coins.u64())) * d
```

Comment:
I suggest to create `Decimal256::from_integer(n_coins)` only once and then reuse it.